velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,251 @@
+namespace {
+
+template <typename TResult, typename TInput>
+class DecimalRoundFunction : public exec::VectorFunction {
+ public:
+  DecimalRoundFunction(
+      int32_t scale,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resultPrecision,
+      uint8_t resultScale)
+      : scale_(
+            scale >= 0
+                ? std::min(scale, (int32_t)LongDecimalType::kMaxPrecision)
+                : std::max(scale, -(int32_t)LongDecimalType::kMaxPrecision)),
+        inputPrecision_(inputPrecision),
+        inputScale_(inputScale),
+        resultPrecision_(resultPrecision),
+        resultScale_(resultScale) {
+    const auto [p, s] = DecimalRoundCallToSpecialForm::getResultPrecisionScale(
+        inputPrecision, inputScale, scale);
+    VELOX_USER_CHECK_EQ(
+        p,
+        resultPrecision,
+        "The result precision of decimal_round is inconsistent with Spark expected.");
+    VELOX_USER_CHECK_EQ(
+        s,
+        resultScale,
+        "The result scale of decimal_round is inconsistent with Spark expected.");
+
+    // Decide the rescale factor of divide and multiply when rounding to a
+    // negative scale.
+    auto rescaleFactor = [&](int32_t rescale) {
+      VELOX_USER_CHECK_GT(
+          rescale, 0, "A non-negative rescale value is expected.");
+      return DecimalUtil::kPowersOfTen[std::min(
+          rescale, (int32_t)LongDecimalType::kMaxPrecision)];
+    };
+    if (scale_ < 0) {
+      divideFactor_ = rescaleFactor(inputScale_ - scale_);
+      multiplyFactor_ = rescaleFactor(-scale_);
+    }
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    auto rawResults =
+        result->asUnchecked<FlatVector<TResult>>()->mutableRawValues();
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vector.
+      applyConstant(rows, args[0], context, rawResults);
+    } else {
+      // Fast path for flat vector.
+      applyFlat(rows, args[0], context, rawResults);
+    }
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+ private:
+  inline TResult applyRound(const TInput input) const {
+    if (scale_ >= 0) {
+      const auto rescaledValue =
+          DecimalUtil::rescaleWithRoundUp<TInput, TResult>(
+              input,
+              inputPrecision_,
+              inputScale_,
+              resultPrecision_,
+              resultScale_);
+      VELOX_DCHECK(rescaledValue.has_value());
+      return rescaledValue.value();
+    } else {
+      VELOX_USER_CHECK(
+          divideFactor_.has_value(),
```

Comment:
Do we need these checks per-row? It feels like we can remove these.

Reply:
Yes, they can be removed.